go/store/blobstore/oci.go

```diff
@@ -0,0 +1,796 @@
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"math"
+	"os"
+	"path"
+	"sort"
+
+	"github.com/oracle/oci-go-sdk/v65/common"
+	"github.com/oracle/oci-go-sdk/v65/objectstorage"
+	"golang.org/x/sync/errgroup"
+)
+
+// 10MB part size
+const minPartSize = 10 * 1024 * 1024
+const defaultPartSize = 5 * minPartSize
+const maxPartNum = 10000
+const defaultBatchSize = 500 * 1024 * 1024
+const defaultConcurrentListeners = 5
+
+type toUpload struct {
+	b       []byte
+	partNum int
+}
+
+type toDownload struct {
+	path    string
+	partNum int
+	br      BlobRange
+}
+
+type uploadFunc func(ctx context.Context, objectName, uploadID string, partNumber int, contentLength int64, reader io.Reader) (objectstorage.CommitMultipartUploadPartDetails, error)
+
+type downloadFunc func(ctx context.Context, objectName, namespace, etag string, partNumber int, br BlobRange) (*toDownload, error)
+
+type tempLocalObject struct {
+	path string
+	f    *os.File
+}
+
+var _ io.ReadCloser = &tempLocalObject{}
+
+func (t *tempLocalObject) Read(p []byte) (int, error) {
+	return t.f.Read(p)
+}
+
+func (t *tempLocalObject) Close() error {
+	err := t.f.Close()
+	os.Remove(t.path)
+	return err
+}
+
+// OCIBlobstore provides an OCI implementation of the Blobstore interface
+type OCIBlobstore struct {
+	provider            common.ConfigurationProvider
+	client              objectstorage.ObjectStorageClient
+	bucketName          string
+	nameSpace           string
+	prefix              string
+	concurrentListeners int
+}
+
+var _ Blobstore = &OCIBlobstore{}
+
+// NewOCIBlobstore creates a new instance of a OCIBlobstore
+func NewOCIBlobstore(provider common.ConfigurationProvider, client objectstorage.ObjectStorageClient, bucketName, prefix string) *OCIBlobstore {
+	for len(prefix) > 0 && prefix[0] == '/' {
+		prefix = prefix[1:]
+	}
+	return &OCIBlobstore{provider, client, bucketName, "", prefix, defaultConcurrentListeners}
+}
+
+func (bs *OCIBlobstore) Path() string {
+	return path.Join(bs.bucketName, bs.prefix)
+}
+
+func (bs *OCIBlobstore) getNamespace(ctx context.Context) (string, error) {
+	if bs.nameSpace != "" {
+		return bs.nameSpace, nil
+	}
+	request := objectstorage.GetNamespaceRequest{}
+	r, err := bs.client.GetNamespace(ctx, request)
+	if err != nil {
+		return "", err
+	}
+	bs.nameSpace = *r.Value
+	return bs.nameSpace, nil
+}
+
+// Exists returns true if a blob exists for the given key, and false if it does not.
+// For InMemoryBlobstore instances error should never be returned (though other
+// implementations of this interface can)
+func (bs *OCIBlobstore) Exists(ctx context.Context, key string) (bool, error) {
+	absKey := path.Join(bs.prefix, key)
+	namespace, err := bs.getNamespace(ctx)
+	if err != nil {
+		return false, err
+	}
+	_, err = bs.client.HeadObject(ctx, objectstorage.HeadObjectRequest{
+		NamespaceName: &namespace,
+		BucketName:    &bs.bucketName,
+		ObjectName:    &absKey,
+	})
+	if err == nil {
+		return true, nil
+	}
+	return false, nil
```

Comment:
This seems a little disingenuous. Seems like we should do the dance:

```go
	if serr, ok := common.IsServiceError(err); ok {
		// handle not found code
		if err.GetHTTPStatusCode() == 404 {
			return false, nil
		}
	}
	return false, err
```

or something like that...